### PR TITLE
[Fixes JIRA #GEOS-8649] Importer does not clean temporary folders aft…

### PR DESCRIPTION
**Pull request by afabiani**
…er imports

### CODE REVIEW

File: src/extension/importer/core/src/main/java/org/geoserver/importer/Directory.java [hunk 1]
```diff
@@ -39,7 +39,7 @@
 public class Directory extends FileData {
 
     private static final Logger LOGGER = Logging.getLogger(Directory.class);
-    
+
```
Comment (aaime) [inline src/extension/importer/core/src/main/java/org/geoserver/importer/Directory.java]: In this file I only see reformats, but might be wrong. Formats changes are not encouraged, and if you have to do them for any reason, place them in a different commit.

File: src/extension/importer/core/src/main/java/org/geoserver/importer/ImportContext.java [hunk 2]
```diff
@@ -278,12 +279,48 @@ public void updated() {
...
+                    }
+                }
+            }
+        }
+        
+        return false;
+    }
+
+    protected boolean canBeRemoved(ImportContext context) {
+        if (context.getState() == ImportContext.State.COMPLETE ||
```
Comment (aaime) [inline src/extension/importer/core/src/main/java/org/geoserver/importer/ImportContext.java]: The formatting is wrong in various places in this method. Also, ifs always need { and } around their blocks (even if one lineers)

File: src/extension/importer/core/src/main/java/org/geoserver/importer/ImportContext.java [hunk 3]
```diff
@@ -278,12 +279,48 @@ public void updated() {
...
+        return false;
+    }
+
+    protected boolean canBeRemoved(ImportContext context) {
+        if (context.getState() == ImportContext.State.COMPLETE || 
+                context.getState() == ImportContext.State.INIT_ERROR)
+            return true;
+        else if (context.getState() == ImportContext.State.PENDING) {
+            for(ImportTask task : context.getTasks()) {
+                if (task.getState() == ImportTask.State.ERROR || task.getError() != null)
```
Comment (aaime) [inline src/extension/importer/core/src/main/java/org/geoserver/importer/ImportContext.java]: Even an errored task can be updated and re-run, imho the files should not be removed. 
However, this might mean some files are not going to to cleaned up in case of shutdown without cleaning the completed and errored imports. Hum... there could be a shutdown hook that cleans the errored ones in case the memory storage is used, maybe? A non clean shutdown will still leave stuff around of course. Thinking out loud, maybe also add a marker file that indicates incomplete imports? (could be used for pending ones too, but not sure)

File: src/extension/importer/core/src/main/java/org/geoserver/importer/ImportContext.java [hunk 4]
```diff
@@ -278,12 +279,48 @@ public void updated() {
...
+                    return true;
+            }
         }
+        return false;
     }
 
+    protected boolean canBePruned(ImportTask task) {
+        if (!task.isDirect() ||
+                task.getState() == ImportTask.State.ERROR || task.getError() != null)
```
Comment (aaime) [inline src/extension/importer/core/src/main/java/org/geoserver/importer/ImportContext.java]: Same as above

File: src/extension/importer/core/src/main/java/org/geoserver/importer/job/JobQueue.java [hunk 5]
```diff
@@ -60,8 +71,19 @@ protected void afterExecute(Runnable r, Throwable t) {
             public void run() {
                 List<Long> toremove = new ArrayList<Long>();
                 for (Map.Entry<Long, Task<?>> e : jobs.entrySet()) {
-                    if (e.getValue().isCancelled() || (e.getValue().isDone() && e.getValue().isRecieved())) {
-                        toremove.add(e.getKey());
+                    if (e.getValue().isCancelled() || (e.getValue().isDone())) {
```
Comment (aaime) [inline src/extension/importer/core/src/main/java/org/geoserver/importer/job/JobQueue.java]: Why the change in logic condition here?

File: src/extension/importer/core/src/main/java/org/geoserver/importer/job/JobQueue.java [hunk 6]
```diff
@@ -60,8 +71,19 @@ protected void afterExecute(Runnable r, Throwable t) {
...
+                        try {
+                            ImportContext context = (ImportContext) e.getValue().get();
+                            if(context.delete()) {
```
Comment (aaime) [inline src/extension/importer/core/src/main/java/org/geoserver/importer/job/JobQueue.java]: Not sure this is "valid", what if the task is done but the context is still running (there can be imports with many tasks, hundreds or thousands)?

File: src/extension/importer/core/src/test/java/org/geoserver/importer/ImporterDataTest.java [hunk 7]
```diff
@@ -431,6 +435,10 @@ public void testImportIntoDatabase() throws Exception {
 
         runChecks("archsites");
         runChecks("bugsites");
+        
+        // Test that temporary data is removed also
+        context.delete();
```
Comment (aaime) [inline src/extension/importer/core/src/test/java/org/geoserver/importer/ImporterDataTest.java]: If you're calling delete manually, you're testing the delete cleans up, but not if the delete will be called by the importer machinery automatically. Probably best to add a REST test and do the checks there too, without a manual delete call?
The JobQueue cleanup frequency should be made configurable, and I guess more frequent than "one minute" per loop (or the test will be too long).

**Comment by aaime** [comment]
@afabiani one more thing, I'm working on a pull request that will make it possible to run executable scripts placed in "importer/scripts". Does it make sense to change this one in order to obtain the following complete directory layout:

````
  data_dir
     - importer
          - importer.properties
          - uploads
          - scripts
````

That is, everything in the same folder? That would match more or less the monitoring module approach.

**Comment by afabiani** [comment]
@aaime I'm going to close this one and re-issue a new PR related to this Issue along with:

1. Improved logic (as per your review's suggestions)
2. Alignment with directory structure
